### PR TITLE
Various debugging utils developed during state machine bug investigation

### PR DESCRIPTION
**Pull request by Spikhalskiy**
These changes are taken from the branch related to investigation and fixes of #648 to keep that branch and future PR clean and easy to follow.
This PR includes various helper logging and functions that could be helpful with state machine debugging.
Including:
- improved states transition logging
- included optional entityName into state machines for concise debug prints
- methods helping with getting a workflow history 
- methods helping with replaying workflow history in tests using TestRule

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/internal/statemachines/CancellableCommand.java [hunk 1]
```diff
@@ -67,7 +67,7 @@ public void handleCommand(CommandType commandType) {
   public WorkflowStateMachines.HandleEventStatus handleEvent(
       HistoryEvent event, boolean hasNextEvent) {
     if (canceled) {
-      return WorkflowStateMachines.HandleEventStatus.NON_MATCHING_EVENT;
+      return WorkflowStateMachines.HandleEventStatus.COMMAND_CANCELLED;
     }
```
Comment (Sushisource) [inline temporal-sdk/src/main/java/io/temporal/internal/statemachines/CancellableCommand.java:71]: Should this change be in this branch?

File: temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java [hunk 2]
```diff
@@ -801,10 +802,10 @@ private void assertMatch(Command command, HistoryEvent event, Object expected, O
     public void workflowTaskStarted(
         long startedEventId, long currentTimeMillis, boolean nonProcessedWorkflowTask) {
       setCurrentTimeMillis(currentTimeMillis);
-      // If some new commands are pending and there are no more command events.
       for (CancellableCommand cancellableCommand : commands) {
+        // TODO delete this block
         if (cancellableCommand == null) {
-          break;
+          throw new Error("What the heck");
         }
```
Comment (Sushisource) [inline temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java:809]: Looks like this guy should be removed :)

File: temporal-sdk/src/main/java/io/temporal/internal/statemachines/CancellableCommand.java (hunk 1 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/statemachines/CancellableCommand.java:71]: No, unrelated, and needed more for clarity and documentation. This change shouldn't change anything.
It's just incorrect to say that the event is not matching when the command is canceled and we never even checked it.
Also, with the current code this path should never actually happen.

File: temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowTaskStateMachine.java [hunk 3]
```diff
@@ -52,7 +53,9 @@ void workflowTaskStarted(
   private final long workflowTaskStartedEventId;
   private final Listener listener;
 
-  private long currentTimeMillis;
+  private long eventTimeOfTheLastWorkflowStartTask;
+  // TODO write a comment describing the difference between workflowTaskStartedEventId and
+  // startedEventId
```
Comment (Sushisource) [inline temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowTaskStateMachine.java:58]: Another TODO probably worth resolving in this PR

Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowTaskStateMachine.java:58]: I don't want to tackle it right now. I don't know the difference and I don't want to dig into this right now. That why I just throw a TODO for a better time.

File: temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java (hunk 2 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java:809]: Yeah, if all tests pass I will throw this away, this thing doesn't make much sense.

**Review by Sushisource — APPROVED** [review]

**Pull request merged**